controller/src/lib.rs

```diff
@@ -2,10 +2,605 @@
+                debug!(log, "outgoing response: {response:?}");
+            }
+
+            // Poll for incoming packets.
+            res = socket.recv_from(&mut recv_buf) => {
+                let (n_bytes, peer) = match res {
+                    Err(e) => {
+                        error!(log, "I/O error receiving UDP packet: {e:?}");
+                        continue;
+                    }
+                    Ok((n_bytes, peer)) => (n_bytes, peer),
+                };
+
+                // Deserialize the message itself.
+                let (message, remainder): (Message, _) = match hubpack::deserialize(&recv_buf) {
+                    Err(e) => {
+                        error!(
+                            log,
+                            "failed to deserialize message";
+                            "reason" => ?e,
+                            "peer" => peer,
+                            "n_bytes" => n_bytes,
+                        );
+                        continue;
+                    }
+                    Ok((msg, remainder)) => (msg, remainder),
+                };
+                debug!(log, "message from peer"; "peer" => peer, "message" => ?message);
+
+                // Sanity check the protocol version.
+                if message.header.version != message::version::V1 {
+                    if let Err(e) = send_version_mismatch(&log, &socket, &peer, &message.header, &message.modules).await {
+                        error!(
+                            log,
+                            "failed to send version mismatch";
+                            "reason" => ?e,
+                            "peer" => peer
+                        );
+                    }
+                    continue;
+                }
+
+                // Sanity check that the message could possibly be meant for us.
+                //
+                // We never expect these message types to be sent to us.
+                if matches!(message.body, MessageBody::HostRequest(_) | MessageBody::HostResponse(_)) {
+                    warn!(log, "wrong message type"; "peer" => peer);
+                    send_protocol_error(&log, &socket, &peer, &message.header, &message.modules).await;
+                    continue;
+                }
+
+                // Check that we have data, if the message is supposed to
+                // contain it.
+                let expected_len = message.expected_data_len();
+                let data = if expected_len > 0  {
+                    if remainder.len() < expected_len {
+                        warn!(
+                            log,
+                            "message did not contain expected data";
+                            "expected_len" => expected_len,
+                            "actual_len" => remainder.len(),
+                            "peer" => peer,
+                        );
+                        send_protocol_error(&log, &socket, &peer, &message.header, &message.modules).await;
```

Comment:
Same question here: if the SP is _replying_ to us, do we need to inform it of this error?

Reply:
I don't know, but it is a clear violation of the protocol. OTOH, it's true that the SP can't do anything except log it, without an update to its software. I'm ambivalent, I guess. One issue I see is, if we don't send a protocol error in this case, why would we ever send one?